backend/lib/input.rb

```diff
@@ -0,0 +1,29 @@
+# frozen_string_literal: true
+
+# Process input parameters
+module Input
+
+  # Return text (null if blank)
+  def self.text(str)
+    str = str.to_s.strip
+    str = nil if str == ""
+
+    return str
+  end
+
+  # Return a number (nil or undefined returns 0)
+  def self.number(n)
+    n = n.to_s.to_i
```

Comment:
That is on purpose. I want to make sure that Input.number *always* returns a number (so anything invalid would return "0" instead of nil).  

I don't mind changing the architecture, but that will mean re-visiting other areas of the code (e.g., allowing the system to return error codes in places where we do not do that), as well as changing existing features (such as auto-correcting bad inputs on things like "update sample type").

Just let me know.

Reply:
So, it is OK to have a method where the caller is responsible for checking that they are giving good input, but the method should be written so that the code is correct.

If you are going to allow for bad input, you need to handle it explicitly and return a result.  At a minimum, this could be returning `nil`, but throwing an exception is also appropriate.

If the API is not giving feedback on malformed input, it needs to. In these cases, the best strategy is to have the code parsing the input throw an exception when the input doesn't match the expected, and then have the controller catch the exception and then give an error response.

Reply:
Hmm, I agree and disagree.  On things like "a password need to have <n> characters and include ..." <or> "a phone number needs to be n-nnn-nnn-nnnn" I agree with "malformed" input.  

But for something that needs to be an integer because I am looking up an index in the DB I traditionally just use "<input>.to_i" (the input is ALWAYS a <string> and has to be converted to an integer regardless).  And ".to_i" will already convert things like "1abc" to "1" and "abc" to "0" and nil to "0".  The only thing my ".to_s.to_i" does is to also convert an <undefined> to "0".  And "0" is NEVER used as an index so that will "fail" gracefully on the subsequent SQL query (by returning nothing.)

Are you OK with the method "input.index" instead of "input.number" which returns "0" when the "index" is "bad"?  That is my strong suggestion. Otherwise I will have to completely change what gets returned to the front-end and there will be huge ramifications.

Reply:
At a minimum change the name of the method.

let's look at the error handling. the api shouldn't just ignore bad input